src/field.rs

```diff
@@ -190,6 +196,55 @@ pub trait FieldElement:
         }
         Ok(vec)
     }
+
+    /// Encode `input` as `bits`-bit vector of elements of `Self` if it's small enough
+    /// to be represented with that many bits.
+    ///
+    /// # Arguments
+    ///
+    /// * `input` - The field element to encode
+    /// * `bits` - The number of bits to use for the encoding
+    /// * `max_input` - The maximal number representable with `bits` bits. Equal to `bits^2-1`
```

Comment:
`max_input` is used as a function parameter because precomputing it is convenient. If chosen incorrectly (i.e., if `max_input < 2^bits - 1`, then this method will panic in a way that the user may not expect. This was fine when this method was crate-private, but I'm a bit wary of putting this in the public API.

The fundamental goal here is to ensure that `input` can be encoded with `bits` bits. I wonder if we can rewrite the function to enforce this some other way? One straightforward approach is to compute `max_input` ourselves, but we can probably do even better in terms of computational efficiency.

Reply:
Great work!